src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -1363,14 +1363,16 @@ object LazyList extends SeqFactory[LazyList] {
 
     private[this] def readObject(in: ObjectInputStream): Unit = {
       in.defaultReadObject()
-      val init = new ArrayBuffer[A]
+      val init = new mutable.ListBuffer[A]
       var initRead = false
       while (!initRead) in.readObject match {
         case SerializeEnd => initRead = true
         case a => init += a.asInstanceOf[A]
       }
       val tail = in.readObject().asInstanceOf[LazyList[A]]
-      coll = init ++: tail
+      // scala/scala#10118: caution that no code path can evaluate `tail.state`
+      // before the resulting LazyList is returned
+      coll = newLL(stateFromIteratorConcatSuffix(init.toList.iterator)(tail.state))
```

Comment:
now that I've suggested this code, there's some argument to be made for calling `init.toList.iterator` outside of `newLL`, but there's also a quite strong argument that it just doesn't matter